apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableMutateRowsRequestManager.cs

```diff
@@ -0,0 +1,254 @@
+                        break;
+                    }
+                }
+            }
+
+            // There was a problem in the data found in SetStatus(), so fail the RPC.
+            if (_messageIsInvalid)
+            {
+                return ProcessingStatus.INVALID;
+            }
+
+            // This list hold indices of entries to be included in retry request
+            List<int> toRetry = new List<int>();
+            ProcessingStatus processingStatus = ProcessingStatus.SUCCESS;
+
+            // Check the current state to determine the state of the results.
+            // There are three states: OK, Fail, or Partial Retry.
+            for (int i = 0; i < Results.Length; i++)
+            {
+                Rpc.Status status = Results[i];
+                if (status.Code == GetGrpcCode(Grpc.Core.StatusCode.OK))
+                {
+                    continue;
+                }
+                if (_rertriableCodes.Contains(status.Code))
+                {
+                    // An individual mutation failed with a retryable code, usually DEADLINE_EXCEEDED.
+                    toRetry.Add(i);
+                    if (processingStatus == ProcessingStatus.SUCCESS)
+                    {
+                        processingStatus = ProcessingStatus.RETRYABLE;
+                    }
+                }
+                else
+                {
+                    // Don't retry if even a single response is not retryable.
+                    processingStatus = ProcessingStatus.NOT_RETRYABLE;
+                    break;
+                }
+            }
+
+            if (toRetry.Count > 0)
+            {
+                _modifiedRequest = CreateRetryRequest(toRetry);
+            }
+            return processingStatus;
+        }
+
+        /// <summary>
+        /// Returns modified MutateRowsRequest.
+        /// </summary>
+        /// <returns></returns>
+        public MutateRowsRequest GetRetryRequest()
+        {
+            return _modifiedRequest;
+        }
+
+        /// <summary>
+        /// Creates a new MutateRowsRequest that's a subset of the original request that
+        /// corresponds to a set of indices.
+        /// </summary>
+        /// <param name="indiciesToRetry">
+        /// LIst of ints that represents indicies of entries that failed.
+        /// </param>
+        /// <returns>
+        /// Returns modified <see cref="MutateRowRequest"/>.
+        /// </returns>
+        private MutateRowsRequest CreateRetryRequest(List<int> indiciesToRetry)
+        {
+            _modifiedRequest = new MutateRowsRequest();
+            _mapToOriginalIndex = new int[indiciesToRetry.Count];
+            for (int i = 0; i < _mapToOriginalIndex.Length; i++)
+            {
+                _mapToOriginalIndex[i] = indiciesToRetry[i];
+                _modifiedRequest.Entries.Add(_originalRequest.Entries[indiciesToRetry[i]]);
+            }
+            return _modifiedRequest;
+        }
```

Comment:
This can be simplified to:
```csharp
private MutateRowsRequest CreateRetryRequest(List<int> indiciesToRetry)
{
    _mapToOriginalIndex = indiciesToRetry.ToArray();
    _modifiedRequest = new MutateRowsRequest { Entries = { indiciesToRetry.Select(i => _originalRequest.Entries[i]) } };
    return _modifiedRequest;
}
```

Reply:
Wouldn't that be slower, since that will traverse the `List` twice?

Reply:
Yes, but I wouldn't try to over-optimize though just yet. I would assume the retries are the exception, not the rule, so you wouldn't have a huge number of them to deal with. But if you are really worried about it, what you have is ok, or you can do the somewhat messier version of:
```csharp
private MutateRowsRequest CreateRetryRequest(List<int> indiciesToRetry)
{
    _mapToOriginalIndex = new int[indiciesToRetry.Count];
    _modifiedRequest = new MutateRowsRequest { Entries =
    {
        indiciesToRetry.Select((indexValue, i) =>
        {
            _mapToOriginalIndex[i] = indexValue;
            return _originalRequest.Entries[indexValue];
        })
    } };
    return _modifiedRequest;
}
```
But I think if you like to go with the optimized version, what you have is better anyway.